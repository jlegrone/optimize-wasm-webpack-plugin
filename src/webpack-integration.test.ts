/* eslint-disable import/no-dynamic-require, global-require */
import * as fs from 'fs';
import * as path from 'path';
import * as webpack from 'webpack';
import * as rimraf from 'rimraf';
import * as binaryen from "binaryen";

jest.mock('optimize-wasm-webpack-plugin');

const cases = fs.readdirSync(path.join(__dirname, './cases'));

describe('Webpack Integration Tests', () => {
  cases.forEach((testCase) => {
    it(testCase, (done) => {
      let options: webpack.Configuration = { entry: { index: './index.js' } };
      const testDirectory = path.join(__dirname, 'cases', testCase);
      const outputDirectory = path.join(__dirname, 'test_output', testCase);
      const configFile = path.join(testDirectory, 'webpack.config.js');

      if (fs.existsSync(configFile)) {
        options = require(configFile);
      }

      options.context = testDirectory;
      if (!options.module) options.module = { rules: [] };
      // if (!options.module.loaders) {
      //   options.module.loaders = [{ test: /\.txt$/, loader: OptimizeWasmPlugin.extract('raw-loader') }];
      // }
      if (!options.output) options.output = { filename: '[name].js' };
      if (!options.output.path) options.output.path = outputDirectory;

      options.stats = "verbose"

      // console.log(options);

      rimraf(outputDirectory, () => {
        webpack(options, (err, stats) => {
          if (err) return done(err);
          if (stats.hasErrors()) return done(new Error(stats.toString()));

          fs.readdirSync(outputDirectory).forEach((file: string) => {
            if (file.endsWith(".wasm")) {
              const filePath = path.join(outputDirectory, file);
              const buffer = fs.readFileSync(filePath);
              const mod = binaryen.readBinary(buffer);
              expect(mod.validate()).toEqual(1);
              const text = mod.emitText();
              expect(text).toMatchSnapshot();
            }
          });

          const expectedDirectory = path.join(testDirectory, 'expected');
          fs.readdirSync(expectedDirectory).forEach((file) => {
            const filePath = path.join(expectedDirectory, file);
            const actualPath = path.join(outputDirectory, file);
            expect(readFileOrEmpty(actualPath)).toEqual(readFileOrEmpty(filePath));
            expect(readFileOrEmpty(actualPath)).toMatchSnapshot();
          });

          done();
        });
      });
    });
  });
});

function readFileOrEmpty(path: string) {
  try {
    return fs.readFileSync(path, 'utf-8');
  } catch (e) {
    return '';
  }
}
