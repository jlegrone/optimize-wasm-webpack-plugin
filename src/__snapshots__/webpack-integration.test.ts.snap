// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webpack Integration Tests add 1`] = `
"(module
 (type $0 (func (param i32 i32) (result i32)))
 (type $1 (func))
 (export \\"a\\" (func $0))
 (export \\"b\\" (func $1))
 (func $0 (; 0 ;) (type $0) (param $0 i32) (param $1 i32) (result i32)
  (i32.add
   (get_local $0)
   (get_local $1)
  )
 )
 (func $1 (; 1 ;) (type $1)
  (nop)
 )
)
"
`;

exports[`Webpack Integration Tests add 2`] = `""`;

exports[`Webpack Integration Tests no-wasm 1`] = `""`;

exports[`Webpack Integration Tests unreachable-pops 1`] = `
"(module
 (type $0 (func))
 (export \\"__webpack_init__\\" (func $0))
 (func $0 (; 0 ;) (type $0)
  (nop)
 )
)
"
`;

exports[`Webpack Integration Tests unreachable-pops 2`] = `""`;
